service/matching/version_sets.go

```diff
@@ -186,22 +231,50 @@ func findVersion(data *persistence.VersioningData, buildID string) findVersionRe
 	}
 }
 
-func makeDefaultSet(data *persistence.VersioningData, setIx int) {
+func makeDefaultSet(data *persistence.VersioningData, setIx int, timestamp *clockpb.HybridLogicalClock) {
+	data.DefaultUpdateTimestamp = timestamp
 	if len(data.VersionSets) <= 1 {
 		return
 	}
-	// Move the set to the end and shift all the others down
-	moveMe := data.VersionSets[setIx]
-	copy(data.VersionSets[setIx:], data.VersionSets[setIx+1:])
-	data.VersionSets[len(data.VersionSets)-1] = moveMe
+	if setIx < len(data.VersionSets)-1 {
+		// Move the set to the end and shift all the others down
+		moveMe := data.VersionSets[setIx]
+		copy(data.VersionSets[setIx:], data.VersionSets[setIx+1:])
+		data.VersionSets[len(data.VersionSets)-1] = moveMe
+	}
 }
 
-func makeVersionInSetDefault(data *persistence.VersioningData, setIx, versionIx int) {
+func makeVersionInSetDefault(data *persistence.VersioningData, setIx, versionIx int, timestamp *clockpb.HybridLogicalClock) {
+	data.VersionSets[setIx].DefaultUpdateTimestamp = timestamp
 	setVersions := data.VersionSets[setIx].BuildIds
 	if len(setVersions) <= 1 {
 		return
 	}
-	moveMe := setVersions[versionIx]
-	copy(setVersions[versionIx:], setVersions[versionIx+1:])
-	setVersions[len(setVersions)-1] = moveMe
+	if versionIx < len(setVersions)-1 {
+		// Move the build ID to the end and shift all the others down
+		moveMe := setVersions[versionIx]
+		copy(setVersions[versionIx:], setVersions[versionIx+1:])
+		setVersions[len(setVersions)-1] = moveMe
+	}
+}
+
+func generateHLCTimestamp(clock clockpb.HybridLogicalClock) clockpb.HybridLogicalClock {
+	wallclock := time.Now().UnixMilli()
+	// Ensure time does not move backwards
+	if wallclock < clock.GetWallClock() {
+		wallclock = clock.GetWallClock()
+	}
+	// Ensure timestamp is monotonically increasing
+	if wallclock == clock.GetWallClock() {
+		clock.Version = clock.GetVersion() + 1
+	} else {
+		clock.Version = 0
+		clock.WallClock = wallclock
+	}
+
+	return clockpb.HybridLogicalClock{WallClock: wallclock, Version: clock.Version, ClusterId: clock.ClusterId}
+}
+
+func zeroHLC(clusterID int64) clockpb.HybridLogicalClock {
+	return clockpb.HybridLogicalClock{WallClock: 0, Version: 0, ClusterId: clusterID}
 }
```

Comment:
I'm going to move these into `common` and document them in the next PR.

Reply:
except make a new package, `common` is way too crowded. or maybe `common/clock`

Reply:
next PR.